snapshot.go

```diff
@@ -128,14 +129,36 @@ func (db *DB) snapshot(ctx context.Context, async bool, onSuccess func()) {
 		return
 	}
 
-	tx := db.beginRead()
+	tx, _, commit := db.begin()
 	level.Debug(db.logger).Log(
 		"msg", "starting a new snapshot",
 		"tx", tx,
 	)
 	doSnapshot := func(writeSnapshot func(context.Context, io.Writer) error) {
+		db.Wait(tx - 1) // Wait for all transactions to complete before taking a snapshot.
 		start := time.Now()
 		defer db.snapshotInProgress.Store(false)
+		defer commit()
+		if db.columnStore.enableWAL {
+			// Appending a snapshot record to the WAL is necessary,
```

Comment:
Hmm don't we already do this somewhere else? I guess it was removed when we switched to using a read txn.

Reply:
Yea it was removed, but I'm adding it back in now that the deadlock is gone